### ISSUE TITLE
Extend the FitRes dataclass

### ISSUE THREAD

**Issue by Riccardinho22** [adap/flower#546]
Hi, I'm trying to extend the FitRes dataclass "on the fly" as you can see below:
`fit_res = FitRes(
            parameters=params_prime,
            num_examples=num_examples_train,
            num_examples_ceil=num_examples_train,
            fit_duration=fit_duration,
        )`
        `fit_res.__class__ = make_dataclass('FitResLoss', fields=[('loss', float)], bases=(FitRes,))`
        `fit_res.loss = float(loss)`
        `print(fit_res.loss)`
       ` return fit_res`
But unfortunately when I get the FitRes value from the `def aggregate_fit` on the Strategy implented class (I change the signature type of the `results: List[Tuple[ClientProxy, FitRes]]` to results: `List[Tuple[ClientProxy, FitResLoss]])` the results doens't have the field `loss`.
How can I extend it?

Thanks in advance!

**Comment by pedropgusmao** [comment]
Hi @Riccardinho22 , thanks for your interest in Flower. As you can see in [here](https://github.com/adap/flower/blob/5e86c095ea87f2adee7904f9208bc8b845fcf40f/src/proto/flwr/proto/transport.proto#L59) `FitRes` is an interface for a protobuf message and you would need to change this and regenerate the messages to include your `loss`. However, notice that a few lines [below](https://github.com/adap/flower/blob/5e86c095ea87f2adee7904f9208bc8b845fcf40f/src/proto/flwr/proto/transport.proto#L65) you also have an `EvaluateRes` that contains a variable `loss` intended to be returned by an `evaluate` method as seen, for example, in the `cifar` code for pytorch below. Let me know if this helps. https://github.com/adap/flower/blob/5e86c095ea87f2adee7904f9208bc8b845fcf40f/src/py/flwr_example/pytorch_cifar/client.py#L88

**Comment by Riccardinho22** [comment]
Hi @pedropgusmao thanks for your super quick reply! 
Yes I'm just seeing that is not manageable add the loss as new FitRes field. I saw the evaluate method.
Anyway I need to use the `loss` ( computed w.r.t. the training dataset) 
https://github.com/adap/flower/blob/5e86c095ea87f2adee7904f9208bc8b845fcf40f/src/py/flwr/server/strategy/strategy.py#L49

Thanks again!

**Comment by pedropgusmao** [comment]
Can you provide more details regarding how you want to evaluate the training loss? I assume you have a different training set for each client. Maybe a partition of an original, larger dataset? Do you want to collect individual losses for the participating clients, or do you want to calculate a kind of "global training loss" where you evaluate the loss on the original larger dataset? If you are using all the clients in the system and you want to obtain a training loss for each client after each round, you could use the `evaluate` function with your training set (instead of using the validation set).

**Comment by danieljanes** [comment]
Thanks for reaching out @Riccardinho22 . In addition to what @pedropgusmao said, we do have plans to generalize the return values of both `fit` (#489) and `evaluate` (#490). This would allow you to return (almost) arbitrary values from `fit` and `evaluate` which comes in handy if you have, for example, multiple losses (or any other metric that you want to keep track of and use in your `Strategy`).

**Comment by Riccardinho22** [comment]
Thanks @pedropgusmao, so I fixed at least for now, storing into external files my requirements, which are: get the loss computed w.r.t. the train dataset as well as the gradient still w.r.t. to the train dataset 
@danieljanes That's a huge improvement, that it'll allow me to perform the stuff stated above without any strange things.
Thanks again!
